beard/clustering/preclustering.py

```diff
@@ -0,0 +1,392 @@
+# -*- coding: utf-8 -*-
+#
+# This file is part of Beard.
+# Copyright (C) 2015 CERN.
+#
+# Beard is a free software; you can redistribute it and/or modify it
+# under the terms of the Revised BSD License; see LICENSE file for
+# more details.
+
+"""The algorithm for preclustering.
+
+.. codeauthor:: Mateusz Susik <mateusz.susik@cern.ch>
+
+"""
+
+import numpy as np
+import six
+
+from beard.utils.names import dm_tokenize_name, first_name_initial
+
+
+class Precluster:
+
+    """Representation of a precluster.
+
+    Stores information about all the author names which belong to this
+    precluster.
+    """
+
+    FULL_NAME_MATCH_BONUS = 10
+
+    def __init__(self, tokens):
+        """Create a precluster. Add first name.
+
+        Parameters
+        ----------
+        :param tokens: tuple
+            Tokens created from the author name in form of
+            ((surname tokens), (first name tokens)).
+        """
+        self._content = {tokens[0]: {tokens[1]: 1}}
+
+        # Note that in case where the cluster is created from a tuple of tokens
+        # with multiple surnames, this signature will be counted to
+        # _single_names_variants. This way we can omit dividing by 0.
+        self._single_names_variants = 1
+        self._name = tokens[0][-1]
+
+    def add_signature(self, tokens):
+        """Add a signature to the precluster.
+
+        Parameters
+        ----------
+        :param tokens: tuple
+            Tokens created from the author name in form of
+            ((surname tokens), (first name tokens)).
+        """
+        if tokens[0] in self._content:
+            if tokens[1] in self._content[tokens[0]]:
+                self._content[tokens[0]][tokens[1]] += 1
+            else:
+                self._content[tokens[0]][tokens[1]] = 1
+        else:
+            self._content[tokens[0]] = {tokens[1]: 1}
+
+        if len(tokens[0]) == 1:
+            self._single_names_variants += 1
+
+    def compare_tokens_from_back(self, tokenized_prefix, last_name):
+        """Match a part of the surname with given names in precluster.
+
+        For example, ``Sanchez-Gomez, Juan`` can appear on a signature as
+        ``Gomez, Juan Sanchez``. This function checks if there is a match
+        between a signature like ``Sanchez-Gomez, Juan``, and names inside the
+        precluster.
+
+        Parameters
+        ----------
+        :param tokenized_prefix: tuple
+            Tokens which represent all given names with few first surnames
+            added. In form of a tuple of strings
```

Comment:
I found this unclear. May be by rephrasing for giving a surname example it would be clear